Extend Candid/Macaroon support

monstermunchkin opened the pull request.

stgraber commented on client/connection.go:
```diff
@@ -159,7 +165,7 @@ func ConnectSimpleStreams(url string, args *ConnectionArgs) (ImageServer, error)
 
 // Internal function called by ConnectLXD and ConnectPublicLXD
-func httpsLXD(url string, args *ConnectionArgs) (ContainerServer, error) {
+func httpsLXD(host string, args *ConnectionArgs) (ContainerServer, error) {
```
We've traditionally been using `uri` when the `url` package is used to avoid masking it.

stgraber commented on lxc/main.go:
```diff
@@ -316,6 +316,22 @@ func (c *cmdGlobal) PreRun(cmd *cobra.Command, args []string) error {
+	candidURL := server.Writable().Config["candid.api.url"]
+	if candidURL != nil {
+		c.conf.CandidURL = candidURL.(string)
+	}
```
I don't see how this code can work. What if the user uses something other than the default remote, what if they do operations against multiple remotes?

Also in the case where the user is adding a new remote, they can't actually know the candid URL until after they've authenticated through candid, so I don't think this approach will work at all.

stgraber commented on lxd/api_1.0.go:
```diff
@@ -378,11 +380,10 @@ func doApi10UpdateTriggers(d *Daemon, nodeChanged, clusterChanged map[string]str
 			fallthrough
 		case "maas.api.key":
 			maasChanged = true
+		case "candid.api.expiry":
```
Shouldn't that also set candidChanged then?

monstermunchkin replied:
It does: `fallthrough`

stgraber replied:
Oh right, missed that somehow.

stgraber commented:
This branch should also take care of restricting the list of valid domains on the server side.

stgraber commented:
- `candid.api.expiry` should be renamed to just `candid.expiry`
 - `candid.domains` needs to be added
 - A new API extension needs to be introduced to track the new config keys
 - The new config keys need to be documented in doc/server.md and added to the bash completion profile
 - We need to find a different way to handle the domain cookie thing, we can't require the client to know the URL of the candid server ahead of time, nor should we ever attempt to store that information since LXD can be reconfigured at any time

stgraber commented:
For the domain cookie thing, I wonder if there's some way to set the cookie so that it is sent regardless of the domain, if so, that may be an option.

Otherwise, look for some way to hook into the bakery code so you can set the cookie after it gets the redirect from LXD, setting it for the right domain at that point.

monstermunchkin commented:
Added docs, renamed config keys, and found a way around the cookie thing. I'm still trying to figure out how to implement the domain validation/restriction on the server side.

stgraber commented:
@monstermunchkin Mind capitalizing the two "rename XYZ..." commits, the lack of consistency with the others bothers me :)

stgraber commented:
Actually, those two should just go away and get squashed with whatever commit introduced the original code

stgraber commented on client/connection.go:
```diff
@@ -159,7 +162,7 @@ func ConnectSimpleStreams(url string, args *ConnectionArgs) (ImageServer, error)
 
 // Internal function called by ConnectLXD and ConnectPublicLXD
-func httpsLXD(url string, args *ConnectionArgs) (ContainerServer, error) {
+func httpsLXD(uri string, args *ConnectionArgs) (ContainerServer, error) {
```
This change doesn't really appear needed anymore.

stgraber commented on client/connection.go:
```diff
@@ -168,7 +171,7 @@ func httpsLXD(url string, args *ConnectionArgs) (ContainerServer, error) {
 	server := ProtocolLXD{
 		httpCertificate:  args.TLSServerCert,
-		httpHost:         url,
+		httpHost:         uri,
```
This change doesn't really appear needed anymore.

stgraber commented on doc/server.md:
```diff
@@ -12,6 +12,8 @@ Key                             | Type      | Default   | API extension
 :--                             | :---      | :------   | :------------            | :----------
 backups.compression\_algorithm  | string    | gzip      | backup\_compression      | Compression algorithm to use for new images (bzip2, gzip, lzma, xz or none)
 candid.api.url                  | string    | -         | candid\_authentication   | URL of the the external authentication endpoint using Candid
+candid.expiry                   | integer   | 3600      | candid\_config           | Candid macaroon expiry
```
Probably worth noting that it's in seconds in the description.

stgraber commented on doc/server.md:
```diff
@@ -12,6 +12,8 @@ Key                             | Type      | Default   | API extension
 backups.compression\_algorithm  | string    | gzip      | backup\_compression      | Compression algorithm to use for new images (bzip2, gzip, lzma, xz or none)
 candid.api.url                  | string    | -         | candid\_authentication   | URL of the the external authentication endpoint using Candid
+candid.expiry                   | integer   | 3600      | candid\_config           | Candid macaroon expiry
+candid.domains                  | string    | -         | candid\_config           | Allowed Candid domains
```
Separator should probably be documented too (comma separator with the parser stripping spaces would be consistent with others)

stgraber commented on shared/version/api.go:
```diff
@@ -122,6 +122,7 @@ var APIExtensions = []string{
 	"container_full",
 	"candid_authentication",
 	"backup_compression",
+	"candid_config",
```
This needs to be documented in doc/api-extensions.md

stgraber commented on lxd/api_1.0.go:
```diff
@@ -417,5 +421,17 @@ func doApi10UpdateTriggers(d *Daemon, nodeChanged, clusterChanged map[string]str
+		expiry := clusterConfig.CandidExpiry()
+		domains := clusterConfig.CandidDomains()
+
+		err := d.setupExternalAuthentication(endpoint, expiry, strings.Split(domains, ","))
```
We should allow both "," and ", "

stgraber replied:
Got this fixed locally, will push once the builders are done rebuilding their cache.
